test/prototype/test_quantized_training.py

```diff
@@ -0,0 +1,225 @@
+    def test_int8_linear_training(self, compile, device):
+        torch._dynamo.reset()
+        bsize = 4
+        embed_dim = 32
+        n_classes = 10
+
+        model_fp32 = nn.Sequential(
+            nn.Linear(embed_dim, embed_dim * 2, bias=False),
+            nn.GELU(),
+            nn.Linear(embed_dim * 2, n_classes),
+        ).to(device)
+        model_int8 = copy.deepcopy(model_fp32)
+        quantize_(model_int8, int8_weight_only_quantized_training())
+
+        if compile:
+            model_fp32.compile()
+            model_int8.compile()
+
+        optim_fp32 = AdamW(model_fp32.parameters())
+        optim_int8 = AdamW(model_int8.parameters())
+
+        # prevent segfault with torch.compile()
+        torch.set_float32_matmul_precision("highest")
+
+        for _ in range(5):
+            inputs = torch.randn(bsize, embed_dim, device=device)
+            labels = torch.randint(n_classes, size=(bsize,), device=device)
+            loss_fp32 = F.cross_entropy(model_fp32(inputs), labels)
+            loss_int8 = F.cross_entropy(model_int8(inputs), labels)
+
+            rel_error = abs(loss_int8.item() - loss_fp32.item()) / abs(loss_fp32.item())
+            assert rel_error < 2e-3, rel_error
+
+            loss_fp32.backward()
+            optim_fp32.step()
+            optim_fp32.zero_grad()
+
+            loss_int8.backward()
+            optim_int8.step()
+            optim_int8.zero_grad()
+
+
+class TestFSDP2(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return 2
+
+    @skip_if_lt_x_gpu(2)
+    def test_fsdp2(self):
+        self.run_subtests(
+            {"compile_layer": [False, True]},
+            self._test_fsdp2,
+        )
+
+    def _test_fsdp2(self, compile_layer):
+        import torch.distributed as dist
+        from torch.distributed._composable.fsdp import fully_shard
+        from torch.testing._internal.distributed._tensor.common_dtensor import ModelArgs, Transformer
+
+        torch._dynamo.reset()
+        batch_size = 3
+        vocab_size = 32
+        seq_len = 64
+        model_args = ModelArgs(
+            n_layers=2,
+            n_heads=2,
+            dim=128,
+            vocab_size=vocab_size,
+            max_seq_len=seq_len,
+            dropout_p=0,
+        )
+        torch.manual_seed(42)
+        base_model = Transformer(model_args).cuda()
+        quantize_(base_model, int8_weight_only_quantized_training())
+        fsdp_model = copy.deepcopy(base_model)
+
+        if compile_layer:
+            for layer in base_model.layers:
+                layer.compile()
+
+        for layer in fsdp_model.layers:
+            if compile_layer:
+                layer.compile()
+            fully_shard(layer)
+        fully_shard(fsdp_model)
+
+        base_optim = AdamW(base_model.parameters(), lr=1e-2)
+        fsdp_optim = AdamW(fsdp_model.parameters(), lr=1e-2)
+
+        # prevent segfault with torch.compile()
+        torch.set_float32_matmul_precision("highest")
```

Comment:
interesting

Reply:
Probably fixed with this https://github.com/pytorch/pytorch/pull/133173